src/vm/dllimportcallback.cpp

```diff
@@ -1416,6 +1416,7 @@ VOID UMThunkMarshInfo::RunTimeInit()
     {
         // caller pop
         m_cbRetPop = 0;
+        m_cbActualArgSize += numRegistersUsed * STACK_ELEM_SIZE;
```

Comment:
Ok, so the stub expects the argument in ecx and it has no stack argument. That means that the original m_cbActualArgSize (the value that came from pStubMD->AsDynamicMethodDesc()->GetNativeStackArgSize() in our case) is the correct value for the size of arguments pushed by the native caller and we should not augment it in any way. Also the numRegistersUsed that you've dumped is correct, since the stub has a single argument in ecx.

That means that the problem likely happens only for NativeCallable case and a fix should be made only for that case. I am not sure though if the right fix is to augment the m_cbActualArgSize  only for this case or if the issue is somewhere else and just exposed here. It is possible that the difference is actually not the number of args in registers and it is just a coincidence. That should be  possible to prove by adding one more parameter, e.g. an int, to the callback function signature in the test (and modify the function pointer / native code calling it accordingly). Then you'd see what you get as GetNativeStackArgSize() and numRegistersUsed.